### PR TITLE
#1567 - Cache knowledge base query results

### PR DESCRIPTION
**Pull request by reckart**
**What's in the PR**
- Introduce a cache at the level of the KnowledgeBaseService which caches queries on a per-KB basis
- Invalidate caches if the configuration of the KBs of a project is changed
- Use cache for read-only KBs for most queries
- Introduce settings to configure the KB cache behavior

**How to test manually**
* Use entity linking
* Use the KB page
* To see cache usage, set the log level for the KB service to `TRACE`

**Automatic testing**
* [ ] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java [hunk 1]
```diff
@@ -22,20 +22,23 @@
...
+    implements KnowledgeBaseProperties
 {
     public static final int HARD_MIN_RESULTS = 10;
     
-    private int defaultMaxResults = 1000;
-    private int hardMaxResults = 10000;
+    private int defaultMaxResults = 1_000;
+    private int hardMaxResults = 10_000;
+    private long cacheSize = 100_000;
+    private int cacheExpireDelay = 15;
```
[the remark this answers is not in the document]
Comment (reckart) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java]: Minutes. We could also go for seconds. I wanted to keep it simply by not making the unit configurable. Maybe too simple?

**Comment by reckart** [comment]
I optimized the caching for the editor indeed. There are still some queries which are bypassing the SPARQLQueryBuilder - those are probably the ones which trigger a lot on the KB page.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcklie) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java]: I would add a comment here what the unit is.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 2]
```diff
@@ -1152,12 +1280,95 @@ public void rebuildFullTextIndex(KnowledgeBase aKB) throws Exception
...
         Optional<KnowledgeBase> kb = Optional.empty();
         String repositoryId = aRepositoryId;
         if (repositoryId != null) {
             kb = getKnowledgeBaseById(aProject, aRepositoryId);
         }
         return kb.isPresent() && kb.get().isEnabled();
     }
+    
+    @Override
+    public List<KBHandle> listHandlesCaching(KnowledgeBase aKB, SPARQLQuery aQuery, boolean aAll)
```
[the remark this answers is not in the document]
Comment (reckart) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: For me "cached" would imply that the method only returns cached data, but not fetching fresh data if necessary. I'll leave it as is for the time being.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (reckart) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/config/KnowledgeBasePropertiesImpl.java]: Switched to `Duration` which allows human-readable settings such as `15m` in the `settings.properties` file.

**Pull request merged**